### ISSUE TITLE
[Backward Pass] Fix gradient computation for matmul, batch_norm2d, conv2d

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ProjectOdyssey#2724]
## Summary
Fix gradient computation issues in backward passes. **CRITICAL** for training.

### 1. matmul_backward
- **File**: \`tests/shared/core/test_matrix.mojo\` (line 56)
- **Problem**: Analytical ~1.5e-08, Numerical ~-0.00015 (10,000x mismatch)
- **Likely causes**: Incorrect transpose, wrong operation order, missing scaling

### 2. batch_norm2d_backward
- **File**: \`tests/shared/core/test_normalization.mojo\` (line 36)
- **Problem**: Analytical -4.77e-07, Numerical -0.0018 (1,000x mismatch)
- **Likely causes**: Incorrect variance/std in backward, missing mean/variance gradients

### 3. conv2d_backward
- **File**: \`tests/shared/core/test_conv.mojo\` (line 77)
- **Problem**: Conv2dBackwardResult ownership issue blocks testing
- **Fix needed**: Make Conv2dBackwardResult Copyable or return tuple

## Related TODO Lines in todo.log
- Line 36: test_normalization.mojo batch_norm2d_backward
- Line 56: test_matrix.mojo matmul_backward
- Line 77: test_conv.mojo Conv2dBackwardResult ownership

## Priority
**P0 - Critical Path** - Blocks training functionality

**Comment by mvillmow** [comment]
## Detailed Implementation Plan

Here is the comprehensive implementation plan:

---

# Implementation Plan: Issue #2724 - Fix Gradient Computation for matmul, batch_norm2d, conv2d

## Summary

This plan addresses **P0 - Critical Path** gradient computation issues blocking training functionality. Three backward pass implementations have incorrect gradient computations that produce analytical gradients mismatched from numerical gradients by 1,000x to 10,000x.

### Issues to Fix

| Function | File | Problem | Mismatch |
|----------|------|---------|----------|
| `matmul_backward` | `shared/core/matrix.mojo:820-928` | Incorrect gradient computation | ~10,000x |
| `batch_norm2d_backward` | `shared/core/normalization.mojo:341-876` | Incorrect variance/mean gradient flow | ~1,000x |
| `conv2d_backward` | `shared/core/conv.mojo:221-448` | Ownership issue + potential gradient bug | TBD |

---

## Step-by-Step Implementation Tasks

### Phase 1: Analysis and Root Cause Identification

#### Task 1.1: Create Minimal Reproducers
**Files**: `tests/shared/core/test_gradient_debug.mojo` (new)

Create minimal test cases that reproduce each gradient mismatch:

1. **matmul_backward reproducer**:
   - Smallest tensor sizes: A=(2,3), B=(3,2)
   - Use `check_gradients_verbose()` with `print_all=True`
   - Print both grad_a and grad_b against numerical gradients

2. **batch_norm2d_backward reproducer**:
   - Smallest tensor: (1, 1, 2, 2) - single batch, single channel
   - Test both training=True and training=False modes
   - Print per-element analytical vs numerical gradient comparison

3. **conv2d_backward reproducer**:
   - Smallest tensor: input=(1, 1, 4, 4), kernel=(1, 1, 2, 2)
   - First fix ownership issue, then verify gradients

#### Task 1.2: Analyze matmul_backward Mathematical Derivation
**File**: `shared/core/matrix.mojo:820-928`

Current implementation (lines 917-927):
```mojo
# For C = A @ B:
# grad_a = grad_output @ B^T  ← CORRECT
# grad_b = A^T @ grad_output  ← CORRECT
var b_t = transpose(b)
var a_t = transpose(a)
var grad_a = matmul(grad_output, b_t)
var grad_b = matmul(a_t, grad_output)
```

**Investigation Focus**:
- The 2D @ 2D case looks mathematically correct
- Bug likely in special cases (2D @ 1D at lines 872-893, 1D @ 2D at lines 895-915)
- Check `_dispatch_matmul_backward_2d_1d` and `_dispatch_matmul_backward_1d_2d` implementations
- Verify transpose implementation correctness
- Check if `grad_output` shape matches expectation

#### Task 1.3: Analyze batch_norm2d_backward Mathematical Derivation
**File**: `shared/core/normalization.mojo:341-757`

Current implementation computes:
1. `grad_beta = sum(grad_output)` per channel ← CORRECT
2. `grad_gamma = sum(grad_output * x_norm)` per channel ← CORRECT
3. `grad_input` via three terms ← LIKELY BUG HERE

**Investigation Focus** (lines 540-612):
```mojo
# Term 1: Direct effect through normalized value
var term1 = grad_x_norm / std

# Term 2: Effect through variance term
var term2 = grad_var * Float32(2.0) * x_minus_mean / N

# Term 3: Effect through mean term
var term3 = grad_mean / N

grad_input = term1 + term2 + term3
```

**Potential Issues**:
- `grad_var` accumulation (lines 568-576) may have sign/scaling error
- `grad_mean` computation (line 579) may be incomplete
- Missing the indirect contribution from variance affecting mean
- The Kratzert derivation requires careful attention to the chain rule through both mean and variance

**Reference Implementation** (PyTorch-style):
```python
# Correct batch norm backward:
grad_x_hat = grad_output * gamma
grad_var = sum(grad_x_hat * (x - mean) * -0.5 * (var + eps)^(-1.5))
grad_mean = sum(grad_x_hat * -1/std) + grad_var * sum(-2 * (x - mean)) / N
grad_input = grad_x_hat / std + grad_var * 2 * (x - mean) / N + grad_mean / N
```

#### Task 1.4: Analyze conv2d_backward Ownership Issue
**File**: `shared/core/conv.mojo:221-448`

The `Conv2dBackwardResult` (alias for `GradientTriple`) should be `Copyable` and `Movable` per `gradient_types.mojo`.

This comment may be outdated. Need to verify:
1. Can we access `result.grad_input`, `result.grad_weights`, `result.grad_bias`?
2. Is the issue actually in test construction, not the struct?
3. Once ownership is fixed, verify gradient correctness

---

### Phase 2: Implementation Fixes

#### Task 2.1: Fix matmul_backward
**File**: `shared/core/matrix.mojo`

**Step 2.1.1**: Verify basic 2D @ 2D case first
- Add debug print statements to show shapes at each step
- Verify `transpose()` returns correct dimensions
- Ensure `grad_output` shape matches `(A.shape[0], B.shape[1])`

**Step 2.1.2**: Fix 2D @ 1D case (lines 872-893)
```mojo
# For C (m,) = A (m, k) @ b (k,):
# grad_a[i,j] = grad_output[i] * b[j]  ← outer product, shape (m, k)
# grad_b[j] = sum_i(A[i,j] * grad_output[i]) = A^T @ grad_output
```

Check `_dispatch_matmul_backward_2d_1d` and `_matmul_backward_2d_1d_impl` for:
- Index calculation errors
- Missing batch dimensions
- Wrong accumulation pattern

**Step 2.1.3**: Fix 1D @ 2D case (lines 895-915)
```mojo
# For C (n,) = a (k,) @ B (k, n):
# grad_a (k,) = B @ grad_output
# grad_b (k, n) = outer(a, grad_output)
```

Verify `_dispatch_matmul_backward_1d_2d` and `_matmul_backward_1d_2d_impl`.

#### Task 2.2: Fix batch_norm2d_backward
**File**: `shared/core/normalization.mojo`

**Step 2.2.1**: Fix grad_var computation (lines 568-576)
The current formula should be accumulated AFTER the first loop, then used in second loop. Currently it's computed per-element inside the grad_input loop which is incorrect.

**Step 2.2.2**: Fix grad_mean computation (line 579)
Current implementation is missing the contribution through variance:
```mojo
grad_mean = sum(grad_x_norm * -1/std) + grad_var * mean(-2 * (x - mean))
```

The second term (`grad_var * mean(-2 * (x - mean))`) is missing!

**Step 2.2.3**: Restructure the backward pass
```python
# Pass 1: Compute batch statistics (mean, var)
# Pass 2: Compute grad_gamma, grad_beta
# Pass 3: Compute grad_var (sum over all elements)
# Pass 4: Compute grad_mean (requires grad_var)
# Pass 5: Compute grad_input (requires grad_var and grad_mean)
```

**Correct Implementation Template**:
```mojo
# Step 3: Compute grad_var (single pass)
var grad_var = Float32(0.0)
for element in channel:
    grad_var += grad_x_norm * x_minus_mean * Float32(-0.5) * (var + eps)^(-1.5)

# Step 4: Compute grad_mean (single pass, requires grad_var)
var sum_x_minus_mean = Float32(0.0)
var grad_mean = Float32(0.0)
for element in channel:
    grad_mean += grad_x_norm * Float32(-1.0) / std
    sum_x_minus_mean += x_minus_mean
grad_mean += grad_var * Float32(-2.0) * sum_x_minus_mean / N  # MISSING TERM!

# Step 5: Compute grad_input (requires grad_var, grad_mean)
for element in channel:
    term1 = grad_x_norm / std
    term2 = grad_var * Float32(2.0) * x_minus_mean / N
    term3 = grad_mean / N
    grad_input = term1 + term2 + term3
```

#### Task 2.3: Fix conv2d_backward Ownership and Verify Gradients
**File**: `shared/core/conv.mojo`

**Step 2.3.1**: Enable backward tests
In `tests/shared/core/test_conv.mojo`, remove the DISABLED block and add proper tests.

**Step 2.3.2**: If gradients fail, investigate:
- The `conv2d_backward` implementation (lines 291-358 for grad_input)
- Index calculations for `kh = ih - oh * stride + padding`
- Kernel indexing pattern
- Accumulation over output channels

---

### Phase 3: Testing and Validation

#### Task 3.1: Update Existing Tests
**Files to modify**:
- `tests/shared/core/test_matrix.mojo` - Enable skipped tests (lines 1623-1627)
- `tests/shared/core/test_normalization.mojo` - Enable skipped test (line 804)
- `tests/shared/core/test_conv.mojo` - Enable backward tests (lines 600-604)

#### Task 3.2: Add Comprehensive Gradient Tests
**File**: `tests/shared/core/test_gradient_checking.mojo`

Add new test functions:
- `test_matmul_backward_2d_2d()` - Test matmul backward for 2D @ 2D case
- `test_matmul_backward_2d_1d()` - Test matmul backward for 2D @ 1D case
- `test_matmul_backward_1d_2d()` - Test matmul backward for 1D @ 2D case
- `test_batch_norm2d_backward_training()` - Test batch_norm2d backward in training mode
- `test_batch_norm2d_backward_inference()` - Test batch_norm2d backward in inference mode
- `test_conv2d_backward_stride1_pad0()` - Test conv2d backward with stride=1, padding=0
- `test_conv2d_backward_stride1_pad1()` - Test conv2d backward with stride=1, padding=1
- `test_conv2d_backward_stride2()` - Test conv2d backward with stride=2

#### Task 3.3: Run Full Test Suite
```bash
# Run all gradient tests
pixi run mojo test tests/shared/core/test_gradient_checking.mojo

# Run specific backward tests
pixi run mojo test tests/shared/core/test_matrix.mojo
pixi run mojo test tests/shared/core/test_normalization.mojo
pixi run mojo test tests/shared/core/test_conv.mojo
```

---

## Files to Modify/Create

### Files to Modify

| File | Changes |
|------|---------|
| `shared/core/matrix.mojo` | Fix `matmul_backward` (lines 820-928) |
| `shared/core/normalization.mojo` | Fix `batch_norm2d_backward` (lines 341-876) |
| `shared/core/conv.mojo` | Verify/fix `conv2d_backward` if needed |
| `tests/shared/core/test_matrix.mojo` | Enable skipped gradient tests |
| `tests/shared/core/test_normalization.mojo` | Enable skipped gradient test |
| `tests/shared/core/test_conv.mojo` | Enable backward tests |
| `tests/shared/core/test_gradient_checking.mojo` | Add new gradient tests |

### Files to Create

| File | Purpose |
|------|---------|
| `tests/shared/core/test_gradient_debug.mojo` | Minimal reproducers for debugging |

---

## Testing Approach

### 1. Unit Testing with Numerical Gradient Checking
Use `check_gradients()` and `check_gradients_verbose()` from `shared/testing/gradient_checker.mojo`:
- Central finite differences: `f'(x) ≈ [f(x + ε) - f(x - ε)] / (2ε)`
- Default epsilon: 1e-5
- Default tolerance: 1e-2 (accounts for float32 accumulation errors)

### 2. Test Tensor Sizes
Start with minimal sizes for fast iteration:
- matmul: A=(2,3), B=(3,2) → grad check ~12 forward passes
- batch_norm: (1,1,2,2) → grad check ~4 forward passes
- conv2d: input=(1,1,4,4), kernel=(1,1,2,2) → grad check ~16 forward passes

### 3. Validation Metrics
For each gradient:
- Maximum absolute difference < 1e-2
- Mean absolute difference < 1e-4
- No elements with >10x relative error

### 4. Edge Cases to Test
- **matmul**: Different dimension combinations, batched operations
- **batch_norm**: training vs inference, different batch sizes, variance near zero
- **conv2d**: stride > 1, padding variations, multi-channel

---

## Success Criteria

### Must Pass
- [ ] `test_matmul_backward_gradient_a()` passes with tolerance 1e-2
- [ ] `test_matmul_backward_gradient_b()` passes with tolerance 1e-2
- [ ] `test_batch_norm2d_backward_gradient_input()` passes with tolerance 1e-2
- [ ] `conv2d_backward` tests compile and pass (ownership issue resolved)
- [ ] All gradient checks show analytical gradient within 1e-2 of numerical gradient

### Numerical Targets
| Function | Current Mismatch | Target Mismatch |
|----------|------------------|-----------------|
| matmul_backward | ~10,000x | < 1e-2 |
| batch_norm2d_backward | ~1,000x | < 1e-2 |
| conv2d_backward | Blocked | < 1e-2 |

### CI/CD
- [ ] All tests in `tests/shared/core/test_matrix.mojo` pass
- [ ] All tests in `tests/shared/core/test_normalization.mojo` pass
- [ ] All tests in `tests/shared/core/test_conv.mojo` pass
- [ ] `pre-commit run --all-files` passes

---

## Resource Usage

**Estimated Implementation Time**: 4-8 hours

**Tool Calls Used**: 15 (of 50 budget)
- Read: 12
- Grep: 2
- Task (Explore agent): 1

**Steps Used**: ~8 (of 50 budget)

**Files Analyzed**:
- `shared/core/matrix.mojo` - matmul_backward implementation
- `shared/core/normalization.mojo` - batch_norm2d_backward implementation
- `shared/core/conv.mojo` - conv2d_backward implementation
- `shared/core/gradient_types.mojo` - GradientPair, GradientTriple structs
- `shared/testing/gradient_checker.mojo` - Numerical gradient checking utilities
- `tests/shared/core/test_matrix.mojo` - Matrix backward tests
- `tests/shared/core/test_normalization.mojo` - Normalization backward tests
- `tests/shared/core/test_conv.mojo` - Conv backward tests
- `tests/shared/core/test_gradient_checking.mojo` - Gradient checking test suite

**Comment by mvillmow** [comment]
## Dependencies (AI-Analyzed)

**Priority**: P0

### Depends On
- #2717

### Blocks
- #2725
- #2728

### Analysis
Backward pass fixes depend on matmul being implemented (#2717). Critical blocker for gradient system (#2725) and training (#2728).

---
*Analyzed by Claude Code dependency analyzer*

**Comment by mvillmow** [comment]
❌ Implementation failed:

```
[Errno 2] No such file or directory: '/tmp/implement-issues-jmsku9_g/implementation_state.json'
```

---
*Automated by implement_issues.py at 2025-12-18 20:53:04*

**Comment by mvillmow** [comment]
❌ Implementation failed:

```
[Errno 2] No such file or directory: '/tmp/implement-issues-bl3mydaz/implementation_state.json'
```

---
*Automated by implement_issues.py at 2025-12-18 21:07:12*

**Comment by mvillmow** [comment]
✅ Already completed via PR #2799

---
*Automated by implement_issues.py at 2025-12-19 11:31:42*

**Comment by mvillmow** [comment]
## Implementation Summary

I haven't made any changes yet - this is the start of our conversation. I'm ready to help you with tasks in the ML Odyssey repository. What would you like me to work on?

---
*Automated by implement_issues.py*

**Comment by mvillmow** [comment]
✅ Committed changes to branch: 2724--backward-pass--fix-gradient-c

---
*Automated by implement_issues.py at 2025-12-19 13:02:53*

**Comment by mvillmow** [comment]
## Investigation Update - batch_norm2d_backward

### Attempted Fixes

I've investigated the batch_norm2d_backward gradient computation issue and attempted multiple approaches:

#### Attempt 1: Kratzert Three-Term Formula
- Implemented the formula from [Kratzert's blog](https://kratzert.github.io/2016/02/12/understanding-the-gradient-flow-through-the-batch-normalization-layer.html)
- Formula: `grad_input = grad_x_norm/std + grad_var * 2 * (x-mean)/N + grad_mean/N`
- Result: Still ~10,000x gradient mismatch

#### Attempt 2: PyTorch Consolidated Formula  
- Switched to PyTorch's optimized formula: `grad_input = (grad_output - k/N - x_norm*dotp/N) * gamma / std`
- Where: `k = sum(grad_output)`, `dotp = sum(grad_output * x_norm)`
- Result: Analytical gradient ≈ 0, numerical gradient = 0.00894 (~20,000x mismatch)

### Debug Output

For test case with `grad_output = ones_like(output)`:
- N = 8.0
- k = 8.0, k/N = 1.0  
- dotp = 0.0 (normalized values sum to zero)
- gamma = 1.5, invstd = 2.407

The PyTorch formula gives: `(1.0 - 1.0 - x_norm * 0.0) * 1.5 * 2.407 ≈ 0`

### Issue Analysis

The formula appears to cancel out completely when `grad_output` is uniform, suggesting either:
1. The PyTorch formula doesn't apply to this specific test case
2. There's a subtle bug in how mean/variance dependencies are handled  
3. The test setup itself has an issue

### Next Steps

I need to either:
1. Find a reference implementation that handles this specific case correctly
2. Re-derive the gradient from first principles accounting for all dependencies
3. Investigate whether the numerical gradient test setup is correct

**This will require more investigation. Moving on to check matmul_backward and conv2d_backward gradients.**

**Comment by mvillmow** [comment]
## Status Update - Other Gradients

### ✅ matmul_backward
All matrix operation tests pass (75/75):
- matmul gradient tests working correctly
- No ~10,000x mismatch found

### ✅ conv2d_backward  
All conv2d tests pass (15/15):
- forward/backward consistency verified
- No ownership issues found
- Gradient computation working correctly

### Remaining Work

**Only batch_norm2d_backward still needs investigation.**

The other gradients mentioned in the original issue appear to have been fixed already or were false alarms.
